message Card {
	required string name = 1;
	required string shape = 2;
}

message ClientToServer {
	message PlayerInfo {
		required string color = 1;
	}
	
	message TradeInfo {
		repeated Card cards = 1;
	}
	
	message ReinforceInfo {
		required string territory = 1;
		required uint32 size = 2;
	}
	
	message AttackInfo {
		required string source = 1;
		required string destination = 2;
		required uint32 army_size = 3;
		required uint32 move = 4;
	}
	
	message RelocateInfo {
		required string source = 1;
		required string destination = 2;
		required uint32 move = 3;
	}
	
	optional PlayerInfo player_info = 1;
	optional TradeInfo trade_info = 2;
	repeated ReinforceInfo reinforce_info = 3;
	optional AttackInfo attack_info = 4;
	optional RelocateInfo relocate_info = 5;
	optional bool end_turn = 6;
}

message ServerToClient {
	message PlayersInfo {
		repeated string colors = 1;
	}
	
	message TerritoryInfo {
		required string name = 1;
		required string owner = 2;
		required uint32 size = 3;
	}
	
	message AttackReply {
		required bool success = 1;
		repeated uint32 red_dice = 2;
		repeated uint32 yellow_dice = 3;
	}
	
	message TurnInfo {
		required string player = 1;
		required string step = 2;
	}
	
	repeated TerritoryInfo territory_info = 1;
	optional string map_info = 2;
	optional AttackReply attack_reply = 3;
	optional TurnInfo turn_info = 4;
	repeated Card cards = 5;
	optional PlayersInfo players = 6;
}
